### PR TITLE
Specs parser improvements

### PR DESCRIPTION
**Pull request by cheshirsky**
- modified async mode of specs data collecting
- ability to get last modification data from git

### CODE REVIEW

File: core/file-tree/index.js [hunk 1]
```diff
@@ -176,17 +58,179 @@ var writeDataFile = function (callback) {
...
+
+    Object.keys(info).forEach(function(key) {
+        specDataObject[key] = info[key];
+    });
+
+    specDataObject['id'] = dirname.replace(meta.root, '').replace(meta.app, '');
+    specDataObject['url'] = '/' + specDataObject['id'];
+    specDataObject['fileName'] = meta.basename;
+
+    var thumbPath = path.join(dirname, 'thumbnail.png');
```
Comment (operatino) [inline core/file-tree/index.js]: Make thumbnail filename configurable.

File: core/file-tree/index.js [hunk 2]
```diff
@@ -176,17 +58,179 @@ var writeDataFile = function (callback) {
...
+    //TODO: false should be replaced by undefined or empty string
+    specDataObject['thumbnail'] = fs.existsSync(thumbPath) ? thumbPath.replace(meta.root, '') : false;
+    if (config.getFilesDateFromGit) {
```
Comment (operatino) [inline core/file-tree/index.js]: We should ckeck if git is installed, even if this feature is turned on.

File: core/file-tree/index.js [hunk 3]
```diff
@@ -176,17 +58,179 @@ var writeDataFile = function (callback) {
...
+        tasks.push(function(next) {
+            shell.exec(command, {silent:true}, function(err, out) {
```
Comment (operatino) [inline core/file-tree/index.js]: Please define command right on top of exec.

File: core/file-tree/index.js (hunk 2 above)
Comment (cheshirsky) [inline core/file-tree/index.js]: This feature is switched off by default. If it is switched on but git is not installed we`d get timestamp from filestat. I don't think we need to check it in any specific way. Shell I add a kind of warning or error handling for that?

File: core/file-tree/index.js (hunk 3 above)
Comment (cheshirsky) [inline core/file-tree/index.js]: I`ll move static part of it into config.

File: core/file-tree/index.js (hunk 1 above)
Comment (cheshirsky) [inline core/file-tree/index.js]: Ok, 4 sure.

**Comment by operatino** [comment]
check out the log http://d.pr/i/15pTj

**Comment by operatino** [comment]
Please add perf logging under verbose flag (using new logger, check API modules), and add async indexing in separate process, because process blocks static server and especially first page is loading quite long..

**Comment by operatino** [comment]
And I would like to see few tests as well.

File: core/file-tree/index.js (hunk 2 above)
Comment (operatino) [inline core/file-tree/index.js]: Didn't noticed that, everything is fine then.

**Comment by operatino** [comment]
After app start on root page visit I am getting this error sometimes:

```[2015-02-18 21:49:26.716] [WARN] app - Error:  [SyntaxError: Unexpected end of input]
SyntaxError: Unexpected end of input
    at Object.parse (native)
    at ParseData.getRaw (/Users/operatino/Dropbox/okp/core/api/parseData.js:84:21)
    at parseData.scope (/Users/operatino/Dropbox/okp/core/api/index.js:222:31)
    at Layer.handle [as handle_request] (/Users/operatino/Dropbox/okp/node_modules/express/lib/router/layer.js:76:5)
    at next (/Users/operatino/Dropbox/okp/node_modules/express/lib/router/route.js:100:13)
    at Route.dispatch (/Users/operatino/Dropbox/okp/node_modules/express/lib/router/route.js:81:3)
    at Layer.handle [as handle_request] (/Users/operatino/Dropbox/okp/node_modules/express/lib/router/layer.js:76:5)
    at /Users/operatino/Dropbox/okp/node_modules/express/lib/router/index.js:227:24
    at Function.proto.process_params (/Users/operatino/Dropbox/okp/node_modules/express/lib/router/index.js:305:12)
    at /Users/operatino/Dropbox/okp/node_modules/express/lib/router/index.js:221:12
```

The same error appears after few main page refreshes, API sends error in this case and navigation tree is not working:

```
{"message":"data not found"}
```

File: core/file-tree/data-collector.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
+var path = require('path');
+var parser = require(path.join(__dirname, 'specs-parser'));
```
Comment (operatino) [inline core/file-tree/data-collector.js]: just use `require('../specs-parser');`

**Comment by operatino** [comment]
'debug' log level is still very poluted:

```
node app -l 'TRACE'
```

```
[2015-02-18 22:02:18.939] [DEBUG] app - Pages tree JSON saved to /Users/operatino/Dropbox/okp/core/api/data/pages-tree.json
[2015-02-18 22:02:18.943] [DEBUG] app - Pages tree JSON saved to /Users/operatino/Dropbox/okp/core/api/data/pages-tree.json
[2015-02-18 22:02:18.963] [DEBUG] app - Pages tree JSON saved to /Users/operatino/Dropbox/okp/core/api/data/pages-tree.json
[2015-02-18 22:02:18.980] [DEBUG] app - Pages tree JSON saved to /Users/operatino/Dropbox/okp/core/api/data/pages-tree.json
[2015-02-18 22:02:18.996] [DEBUG] app - Pages tree JSON saved to /Users/operatino/Dropbox/okp/core/api/data/pages-tree.json
[2015-02-18 22:02:19.036] [DEBUG] app - Pages tree JSON saved to /Users/operatino/Dropbox/okp/core/api/data/pages-tree.json
[2015-02-18 22:02:19.076] [DEBUG] app - Pages tree JSON saved to /Users/operatino/Dropbox/okp/core/api/data/pages-tree.json
[2015-02-18 22:02:19.098] [DEBUG] app - Pages tree JSON saved to /Users/operatino/Dropbox/okp/core/api/data/pages-tree.json
```

**Comment by operatino** [comment]
With git dates turned on from time to time, on main page refresh I get undefined data http://d.pr/i/10P1Z

File: core/file-tree/data-collector.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+		parser.parse(message.config.specsRoot, function(specsData) {
+			process.send(specsData);
+		});
+	};
+	
+	this._init = function() {
+		if(msg) {
+			this.dataCollectionTask(msg);
+		} else {
+			console.log("Child process resieved empty message. Unable to start processing.");
```
Comment (operatino) [inline core/file-tree/data-collector.js]: resieved -> received

**Comment by operatino** [comment]
Check linting errors

File: core/file-tree/index.js [hunk 6]
```diff
@@ -1,208 +1,71 @@
...
+    'excludedDirs': ['data', 'plugins', 'node_modules', '.git', '.idea'],
+    'cron': false,
+    'cronProd': true,
+    'cronRepeatTime': 60000,
+    'outputFile': path.join(global.pathToApp, 'core/api/data/pages-tree.json'),
+    'specsRoot': path.join(global.pathToApp, global.opts.core.common.pathToUser).replace(/\\/g, '/'),
+    'getFilesDateFromGit': true,
+    'infoFile': "info.json",
+    'thumbnailFileName': "thumbnail.png",
+    'specFileRegExPattern': "index\.(html|src)",
```
Comment (operatino) [inline core/file-tree/index.js]: I was doing some changes on index file mask, now there's other extensions.

File: core/file-tree/specs-parser.js [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+        specDataObject['id'] = dirname.replace(meta.root, '').replace(meta.app, '');
+        specDataObject['url'] = '/' + specDataObject['id'];
+        specDataObject['fileName'] = meta.basename;
+
+        var thumbPath = path.join(dirname, config.thumbnailFileName);
+        //TODO: false should be replaced by undefined or empty string
+        specDataObject['thumbnail'] = fs.existsSync(thumbPath) ? thumbPath.replace(meta.root, '') : false;
+        if (config.getFilesDateFromGit) {
+            tasks.push(function(next) {
+                shell.exec(config.gitCommandBase + dirname, {silent:true}, function(err, out) {
```
Comment (operatino) [inline core/file-tree/specs-parser.js]: And if shell gets an error? Or git is not installed.

**Comment by operatino** [comment]
I will ask you to describe new indexing process in skype, and please don't forget about perf metrics that I asked. Use `pretty-hrtime` for that, as I did in markdown support branch.

**Comment by operatino** [comment]
Please re-open PR to 0.5.0 branch.

**Pull request closed without merging**